Shared Memory on C++ API

yellowhatter opened the pull request:
SHM functionality implemented on C++ API

Transparently places ROS messages that are above configurable size threshold into shared-memory.

yellowhatter committed:
C++ SHM

yellowhatter committed:
fix default shm msg size threshold

yellowhatter committed:
code format fix

yellowhatter committed:
add lost tracepoints

yellowhatter committed:
code format

yellowhatter committed:
remove unnecessry flags from zenoh-cpp vendor

yellowhatter committed:
review fixes

yellowhatter committed:
codestyle fixes

yellowhatter committed:
Change 2023 -> 2024 in copyright

yellowhatter committed:
Remove obsolete comment from default config

ahcorde commented on rmw_zenoh_cpp/src/detail/shm_context.hpp:
```diff
@@ -0,0 +1,36 @@
+struct ShmContext
+{
+  zenoh::PosixShmProvider shm_provider;
+  size_t msgsize_threshold;
```
include `<cstddef>`

yellowhatter replied:
fixed

ahcorde commented on rmw_zenoh_cpp/src/detail/shm_context.cpp:
```diff
@@ -0,0 +1,33 @@
+namespace rmw_zenoh_cpp
+{
+///=============================================================================
+ShmContext::ShmContext(size_t alloc_size, size_t msgsize_threshold)
```
include `<cstddef>`

yellowhatter replied:
fixed in `shm_context.hpp`

yellowhatter committed:
include <cstddef>

yellowhatter committed:
Merge commit '2429bf27c798af7be0b759bea1d770152b0f5042'

Conflicts:
	rmw_zenoh_cpp/src/detail/rmw_subscription_data.hpp

yellowhatter committed:
style fix

yellowhatter committed:
bugfixed zenoh

imstevenpmwork committed:
fix: Update VCS_VERSION for zenoh_cpp_vendor to latest commit (#52)

yellowhatter committed:
Merge commit '0ef1d11cfa09fb6d409706c48a8e471bb72fef68'

yellowhatter committed:
bump zenoh

yellowhatter committed:
- non-blocking allocation for SHM
- fallback to non-shm allocation if SHM failed

yellowhatter committed:
change defaulkt shm size to 16mb to fit big topics

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:
```diff
@@ -165,21 +165,21 @@ class rmw_context_impl_s::Data final : public std::enable_shared_from_this<Data>
-        throw std::runtime_error("Unable to create shm provider.");
-      }
-      shm_provider_ = std::move(provider);
+#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY
```
I will say that I'm not a huge fan of wrapping this in a `#ifdef` this way.  When we release it on the buildfarm, we can only choose one define here, so we may as well just unconditionally add it.

In other words, if we think this feature is stable enough to use, we should just always build it in.  If it is not yet stable enough to use, then we shouldn't merge the PR until we are confident in it.

yellowhatter replied:
This define is not about the stability but about binary size. SHM is relatively big subsystem of Zenoh with additional dependencies (and it will get even larger in the future) and some (yet small) wire, computation and memory overhead. We have shared-memory feature conditionally compiled through the whole bunch of our projects: zenoh -> zenoh-c -> zenoh-cpp, so supporting this in zenoh_rmw makes sense for users who care about binary size. You can release RMW with `RMW_ZENOH_BUILD_WITH_SHARED_MEMORY=true` (which is `true` by default), but there still will be an option for users to make custom build without SHM.

clalancette replied:
The problem is that most of our users will have no ability to enable/disable this, since they will use the binary packages.  And in a typical ROS 2 installation, my opinion is that this is not going to be the biggest bit; rclcpp, for instance, will be far larger.  So I still feel like we should just unconditionally compile this in.

@Yadunund how do you feel about it?

ruffsl replied:
Would this be for users who are building the stack from source via Yocto for embedded targets? As in, for memory constrained applications that only a respond one ROS node per device?

JEnoch replied:
The `#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY` has been removed in f5520c37dfd55952003f8ba8e55c93d2c9ec75db.
Since #692 the `shared-memory` feature is always enabled when vendoring `zenoh-c` for `rmw_zenoh`.

clalancette commented on rmw_zenoh_cpp/src/detail/zenoh_config.cpp:
```diff
@@ -43,6 +43,14 @@ static const std::unordered_map<ConfigurableEntity,
+static const char * zenoh_shm_enabled_envar = "ZENOH_SHM_ENABLED";
+static const bool zenoh_shm_enabled_default = true;
+static const char * zenoh_shm_alloc_size_envar = "ZENOH_SHM_ALLOC_SIZE";
+static const size_t zenoh_shm_alloc_size_default = 16 * 1024 * 1024;
```
What is the behavior of the shared-memory code if a message is published that is larger than this?  Given my comment above, I think that we should make it fall-back to the "regular" allocation path and not use shared-memory.

yellowhatter replied:
Currently rmw_zenoh falls back to common allocation if for any reason the SHM allocation failed

yellowhatter committed:
Merge commit '3739a0f1bceaed7e923d8f5d25b9db1f24f1d1fc'

Commit:
Merge commit 'f040f5cc2d2a2a4d7c6e8713e9526a6273b95427'

yellowhatter committed:
Merge commit '2930620850d09474b3178518f8fcd66b19fc9967'

yellowhatter committed:
Merge commit 'bda2d7181d7e1ea2f7d143fabf7c4ffc6c670b20'

yellowhatter committed:
remove RMW_ZENOH_BUILD_WITH_SHARED_MEMORY

yellowhatter committed:
up zenoh (optimized SHM subsystem)

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
Merge commit 'e638f8c5ea79d614596ea4dfa89f07dc8065ad93'

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
up zenoh

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
Update CMakeLists.txt

yellowhatter committed:
change default message size threshold fro SHM

yellowhatter committed:
Merge commit '2e43a482ce82a475f4b561f0722bdbe7becd18c3'

Conflicts:
	zenoh_cpp_vendor/CMakeLists.txt

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/rmw-zenoh-binaries-for-rolling-jazzy-and-humble/41395/26

yellowhatter committed:
Merge commit 'bab4c93a8c27a7e686d85c1fd7404f2bd7436912'

Conflicts:
	rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp

yellowhatter committed:
- fix style
- add non-shm fallback to serialized message publication

yellowhatter committed:
fix compilation

yellowhatter committed:
Merge commit '699de4682ca7c2a145315c66d2a52d25196059af'

yellowhatter committed:
Merge commit '76ab162d27997995023c8ea8525da328dc4c4b4b'

yellowhatter committed:
Merge commit '8d6eea5cbf17de11715e929220052257a26dc254'

yellowhatter committed:
Merge commit 'ea63b21ba4d1ca56f4cfe5eb3022204826d5cd75'

yellowhatter committed:
Merge commit 'caaeddef64707682b205d20d75c8454d0ccd232c'

Conflicts:
	rmw_zenoh_cpp/config/DEFAULT_RMW_ZENOH_ROUTER_CONFIG.json5
	rmw_zenoh_cpp/config/DEFAULT_RMW_ZENOH_SESSION_CONFIG.json5
	rmw_zenoh_cpp/src/detail/zenoh_config.cpp

yellowhatter committed:
Merge commit 'a5385965daeacf1c84105abed0b58621413e667c'

yellowhatter committed:
Make shm enabled state read from zenoh config

yellowhatter committed:
Merge commit 'da47bcea37f93aac2ed3ad3ba3c023499830e512'

yellowhatter committed:
up zenoh

mcp-ats commented:
Hi, what is the plan for Jazzy? Will this be backported to Jazzy?

I cannot try right now on rolling but I tried the following branch from Zetta:

https://github.com/ZettaScaleLabs/rmw_zenoh/commits/dev/1.0.0-cpp-shm-jazzy/

It works despite some complaints on the subscription side:

<details><summary>ros2 run image_tools showimage</summary>

```console
>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'failed to resolve symbol 'rmw_event_type_is_supported' in shared library '/home/mcp/ws_zenoh/install/rmw_zenoh_cpp/lib/librmw_zenoh_cpp.so', at ./src/functions.cpp:152'

with this new error message:

  'Expected lexeme type (19) not found, search ended at index 40, at ./src/rcl/lexer_lookahead.c:239'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
```
</details>

<details><summary>ros2 topic hz /image -w 200</summary>

```console
>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'failed to resolve symbol 'rmw_event_type_is_supported' in shared library '/home/mcp/ws_zenoh/install/rmw_zenoh_cpp/lib/librmw_zenoh_cpp.so', at ./src/functions.cpp:152'

with this new error message:

  'service's implementation is invalid, at ./src/rcl/service.c:427'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
```
</details>

I will try to use rolling as soon as I can!

mcp-ats commented:
Hey I tried with rolling and with this PR but I have some sporadic crashes at the beginning of the publisher, see below.

<details><summary>ros2 run image_tools cam2image --ros-args -p burger_mode:=true -p width:=1920 -p height:=1080 -p frequency:=50.0</summary>

```console
2025-04-25T16:08:24.657452Z  WARN Watchdog Confirmator ThreadId(03) zenoh_shm::watchdog::periodic_task: Some("Watchdog Confirmator"): error setting scheduling priority for thread: OS(1), will run with priority 48. This is not an hard error and it can be safely ignored under normal operating conditions. Though the SHM subsystem may experience some timeouts in case of an heavy congested system where this watchdog thread may not be scheduled at the required frequency.
2025-04-25T16:08:24.657463Z  WARN   Watchdog Validator ThreadId(04) zenoh_shm::watchdog::periodic_task: Some("Watchdog Validator"): error setting scheduling priority for thread: OS(1), will run with priority 48. This is not an hard error and it can be safely ignored under normal operating conditions. Though the SHM subsystem may experience some timeouts in case of an heavy congested system where this watchdog thread may not be scheduled at the required frequency.
[INFO] [1745597304.739037321] [cam2image]: Publishing image #1
[INFO] [1745597304.757779241] [cam2image]: Publishing image #2
[INFO] [1745597304.777845078] [cam2image]: Publishing image #3
[INFO] [1745597304.797739350] [cam2image]: Publishing image #4
[INFO] [1745597304.817716038] [cam2image]: Publishing image #5
[INFO] [1745597304.837761351] [cam2image]: Publishing image #6
[INFO] [1745597304.857756956] [cam2image]: Publishing image #7
[INFO] [1745597304.877738982] [cam2image]: Publishing image #8
[INFO] [1745597304.897754033] [cam2image]: Publishing image #9
[INFO] [1745597304.917714415] [cam2image]: Publishing image #10
thread '<unnamed>' panicked at library/std/src/time.rs:417:33:
overflow when adding duration to instant
stack backtrace:
   0:     0x77697b7ef1cc - std::backtrace_rs::backtrace::libunwind::trace::he991e66e5f22a1db
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x77697b7ef1cc - std::backtrace_rs::backtrace::trace_unsynchronized::h43f35a5576eecb13
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x77697b7ef1cc - std::sys_common::backtrace::_print_fmt::hf9c799a68b2477fc
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x77697b7ef1cc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf2d48dfeb5948660
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x77697b6acb80 - core::fmt::rt::Argument::fmt::h507ca91fbb1ef494
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/fmt/rt.rs:142:9
   5:     0x77697b6acb80 - core::fmt::write::h4d5f6025aa566322
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/fmt/mod.rs:1120:17
   6:     0x77697b7e69a1 - std::io::Write::write_fmt::hc088b3c8cc6dcb9f
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/io/mod.rs:1762:15
   7:     0x77697b7eef6e - std::sys_common::backtrace::_print::h231201f6cff8f56a
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x77697b7eef6e - std::sys_common::backtrace::print::h478d6be486f8cd2d
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x77697b80c640 - std::panicking::default_hook::{{closure}}::hf1c55106b4d45a90
  10:     0x77697b80d190 - std::panicking::default_hook::h7b29000c631cc719
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/panicking.rs:292:9
  11:     0x77697b80d190 - std::panicking::rust_panic_with_hook::h2a988919dd121561
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/panicking.rs:779:13
  12:     0x77697b7ef76c - std::panicking::begin_panic_handler::{{closure}}::h3a9eaa1d23ecf555
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/panicking.rs:657:13
  13:     0x77697b7ef6c6 - std::sys_common::backtrace::__rust_end_short_backtrace::h813561fdec15f700
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/sys_common/backtrace.rs:170:18
  14:     0x77697b80c9c1 - rust_begin_unwind
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/std/src/panicking.rs:645:5
  15:     0x77697b4c2cf4 - core::panicking::panic_fmt::hef0307862026e6f9
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/panicking.rs:72:14
  16:     0x77697b4c3382 - core::panicking::panic_display::hed86ea0466f0a3dc
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/panicking.rs:178:5
  17:     0x77697b4c3382 - core::panicking::panic_str::h1dc008f642ffb82c
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/panicking.rs:152:5
  18:     0x77697b4c3382 - core::option::expect_failed::hd0fb791ee88ae8a9
                               at /build/rustc-ntAYxy/rustc-1.75.0+dfsg0ubuntu1/library/core/src/option.rs:1985:5
  19:     0x77697bc2d699 - zenoh_transport::common::pipeline::StageIn::push_network_message::h170f37ff16d11688
  20:     0x77697bc25acb - zenoh_transport::common::pipeline::TransmissionPipelineProducer::push_network_message::h37560962fda501bc
  21:     0x77697bc22032 - <zenoh_transport::unicast::universal::transport::TransportUnicastUniversal as zenoh_transport::unicast::transport_unicast_inner::TransportUnicastTrait>::schedule::he921124ad80cc08e
  22:     0x77697b8bdd96 - <zenoh::net::primitives::mux::Mux as zenoh::net::primitives::EPrimitives>::send_push::h56fea9c3b8ac4c8f
  23:     0x77697b88685b - zenoh::net::routing::dispatcher::face::Face::send_push_lazy::h2662885c844bbfa9
  24:     0x77697b884ca4 - zenoh::api::session::SessionInner::resolve_put::ha897222f5ca758e8
  25:     0x77697b68db46 - ze_advanced_publisher_put
  26:     0x77698042f6a6 - _ZN13rmw_zenoh_cpp13PublisherData7publishEPKvRKSt8optionalINS_10ShmContextEE
  27:     0x77698045036b - rmw_publish
  28:     0x776981d971a3 - <unknown>
  29:     0x77697bfa4e26 - <unknown>
  30:     0x77697bfac68f - <unknown>
  31:     0x77697bfbe1b5 - <unknown>
  32:     0x776981f0c5e1 - _ZN6rclcpp8Executor22execute_any_executableERNS_13AnyExecutableE
  33:     0x776981f1d564 - _ZN6rclcpp9executors22SingleThreadedExecutor4spinEv
  34:     0x5ead878e2007 - main
  35:     0x77698162a1ca - __libc_start_call_main
                               at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  36:     0x77698162a28b - __libc_start_main_impl
                               at ./csu/../csu/libc-start.c:360:3
  37:     0x5ead878e31c5 - _start
  38:                0x0 - <unknown>
[ros2run]: Aborted
```
</details>

subscribers are not complaining in rolling but at 50.0 Hz it sometimes crashes and if I increase further the frequency I get always a crash.

yellowhatter committed:
Merge commit 'a9ab960d57e1bb5eaa966e3dd90a5b32b5a14b61'

Conflicts:
	zenoh_cpp_vendor/CMakeLists.txt

yellowhatter committed:
Merge commit '2771bc22a2646b8b37382a1379538cdd51a4c4b2'

yellowhatter committed:
Up zenoh (fixes batching wait regression)

Yadunund commented:
@yellowhatter could you resolve conflicts with `main`?

yellowhatter committed:
Merge commit '33b250e4541718977f35ea6d518a96f5af7465fd'

Conflicts:
	zenoh_cpp_vendor/CMakeLists.txt

yellowhatter commented:
> @yellowhatter could you resolve conflicts with `main`?

done

mcp-ats commented:
@yellowhatter I tried to build it but I get the same issue of the pipeline with the lock file.

I updated rustc via rustup to 1.85.0 but I still get an error about the lock file

`lock file version `4` was found, but this version of Cargo does not understand this lock file, perhaps Cargo needs to be updated?`

Any suggestion?

yellowhatter committed:
do not use Zenoh release tags as they are not compatible with 1.75 because of Cargo.lock version

yellowhatter commented:
> @yellowhatter I tried to build it but I get the same issue of the pipeline with the lock file.
> 
> I updated rustc via rustup to 1.85.0 but I still get an error about the lock file
> 
> `lock file version `4` was found, but this version of Cargo does not understand this lock file, perhaps Cargo needs to be updated?`
> 
> Any suggestion?

hello @mcp-ats ! That's my fault sorry. Fixed

mcp-ats commented:
thank you @yellowhatter , I tried it and it seems to be working properly now on rolling.

I quickly tested it with one publisher as cam2image tool, with 4K images at 30Hz and 6 subscribers using topic Hz.

I did not notice any problem now.

One more question, I would like to consider Zenoh where I am working but we are committing to Jazzy. Do you think it might be feasible to backport this to Jazzy or do you foresee critical issues?

Thanks!

ahcorde reviewed (changes requested):
can we add in the README some notes about how to enable/disable this feature?

yellowhatter commented:
> thank you @yellowhatter , I tried it and it seems to be working properly now on rolling.
> 
> I quickly tested it with one publisher as cam2image tool, with 4K images at 30Hz and 6 subscribers using topic Hz.
> 
> I did not notice any problem now.
> 
> One more question, I would like to consider Zenoh where I am working but we are committing to Jazzy. Do you think it might be feasible to backport this to Jazzy or do you foresee critical issues?
> 
> Thanks!

Porting to Jazzy is doable, there are no specific blockers.

yellowhatter committed:
Merge commit '5ea12c461d8a11e62fb8d20bc31fef3ff9f6992d'

ahcorde commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -263,11 +294,11 @@ rmw_ret_t PublisherData::publish(
-  zenoh::Bytes payload(std::move(raw_data));
+  auto payload = shmbuf.has_value() ? zenoh::Bytes(std::move(*shmbuf)) :
+    zenoh::Bytes(
+    std::vector<uint8_t>(
```
include `<cstdint>`

yellowhatter replied:
fixed

yellowhatter commented:
> thank you @yellowhatter , I tried it and it seems to be working properly now on rolling.
> 
> I quickly tested it with one publisher as cam2image tool, with 4K images at 30Hz and 6 subscribers using topic Hz.
> 
> I did not notice any problem now.
> 
> One more question, I would like to consider Zenoh where I am working but we are committing to Jazzy. Do you think it might be feasible to backport this to Jazzy or do you foresee critical issues?
> 
> Thanks!

Pls check this: https://github.com/ros2/rmw_zenoh/pull/656

yellowhatter committed:
Add SHM section to  README.md

yellowhatter committed:
fix list

yellowhatter committed:
Review fix

yellowhatter commented:
> can we add in the README some notes about how to enable/disable this feature?

Added

Yadunund committed:
Resolve merge conflicts with rolling

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund committed:
Move SHMContext into zenoh_utils

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund reviewed (changes requested):
@yellowhatter the implementation looks good but I have one concern with thread-safety. Specifically around the access to and lifetime of `const std::optional<ShmContext> & shm` passed to the `publish()` call.

`context_impl->shm()` locks an internal mutex and returns `const std::optional<ShmContext> & shm` but doesn't guarantee mutually exclusive access to the `ShmContext` by two or more publishers whose `publish()` methods could be called concurrently. Further, if the `context_impl` is destroyed, while `publish()` is holding onto `const std::optional<ShmContext> & shm`, it could lead to UB.

Here are my suggestions

1. Make `ShmContext` a class with data members moved to private scope along with a `mutex`. 
2. Add a method to `ShmContext` that accepts a `char * msg_bytes`, locks a mutex, and attempt to allocate it using the `shm_provider`. Then within `publish()`, call this method while passing `msg_bytes` which was initialised as nullptr. If after calling this method, `msg_bytes` is still `nullptr`, default to the usual allocator. 
3. Store a `shared_ptr<ShmContext` within the `context_impl`. Initialize to `nullptr` if SHM not configured. 
4. Update the constructor of `PublisherData` to accept this `shared_ptr<ShmContext>`.  No need to pass it by reference to the `publish()` call. Within `publish()` if this `ShmContext` is `nullptr`, treat it as SHM was not enabled.

let me know what you think?

milidam commented on README.md:
````diff
@@ -217,9 +217,34 @@ Here are two ways to configure on the remote side:
+```
+
+The following additional configuration options available as environment variables:
+- `ZENOH_SHM_ALLOC_SIZE`: size (in bytes) of memory to allocate as shared memory arena. Must be a multiply of 4. The default value is 16Mb.
````
Shouldn't it be 16MB ("in bytes", "16 * 1024 * 1024")?

```suggestion
- `ZENOH_SHM_ALLOC_SIZE`: size (in bytes) of memory to allocate as shared memory arena. Must be a multiply of 4. The default value is 16MB.
```

yellowhatter committed:
Merge commit '2a1544a7c375497c05e06743a15b75709cdf05ac'

Conflicts:
	rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp

yellowhatter committed:
code format

yellowhatter committed:
style fix

yellowhatter committed:
unused space

yellowhatter committed:
Up zenoh

yellowhatter committed:
- use new SHM API
- fix flaw with SHM allocation blocking

yellowhatter committed:
Update README.md

Co-authored-by: milidam <milidam@users.noreply.github.com>
Signed-off-by: yellowhatter <104833606+yellowhatter@users.noreply.github.com>

yellowhatter commented:
> @yellowhatter the implementation looks good but I have one concern with thread-safety. Specifically around the access to and lifetime of `const std::optional<ShmContext> & shm` passed to the `publish()` call.
> 
> `context_impl->shm()` locks an internal mutex and returns `const std::optional<ShmContext> & shm` but doesn't guarantee mutually exclusive access to the `ShmContext` by two or more publishers whose `publish()` methods could be called concurrently. Further, if the `context_impl` is destroyed, while `publish()` is holding onto `const std::optional<ShmContext> & shm`, it could lead to UB.
> 
> Here are my suggestions
> 
>     1. Make `ShmContext` a class with data members moved to private scope along with a `mutex`.
> 
>     2. Add a method to `ShmContext` that accepts a `char * msg_bytes`, locks a mutex, and attempt to allocate it using the `shm_provider`. Then within `publish()`, call this method while passing `msg_bytes` which was initialised as nullptr. If after calling this method, `msg_bytes` is still `nullptr`, default to the usual allocator.
> 
>     3. Store a `shared_ptr<ShmContext` within the `context_impl`. Initialize to `nullptr` if SHM not configured.
> 
>     4. Update the constructor of `PublisherData` to accept this `shared_ptr<ShmContext>`.  No need to pass it by reference to the `publish()` call. Within `publish()` if this `ShmContext` is `nullptr`, treat it as SHM was not enabled.
> 
> 
> let me know what you think?

Hi @Yadunund , 
With regards to `SHMContext`, the `PosixShmProvider` that is used there is fully threadsafe, so it is absolutely fine to have concurrent allocations.
With regars to `SHMContext` lifetime, yes, let's then put it from `option` to `shared_ptr` and then rely on `shared_ptr` semantic to have proper lifetime management without additional explicit locking.

yellowhatter committed:
Make ShmContext lifetime more robust

yellowhatter committed:
Fix compilation error

YuanYuYuan commented:
Pulls: ros2/rmw_zenoh#363
Gist: https://gist.githubusercontent.com/YuanYuYuan/2b9f57fd08f0f435b9ec0baece14168a/raw/4cc3ae8525eaadd998d577b813e07a322eee8c19/ros2.repos
BUILD args: --packages-above-and-dependencies rmw_zenoh_cpp
TEST args: --packages-above rmw_zenoh_cpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16635
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24495)](http://ci.ros2.org/job/ci_linux/24495/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18551)](http://ci.ros2.org/job/ci_linux-aarch64/18551/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3952)](http://ci.ros2.org/job/ci_linux-rhel/3952/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25045)](http://ci.ros2.org/job/ci_windows/25045/)

yellowhatter committed:
Merge commit '2ca58518d12e0344e15193c6a389b0bb3a108543'

yellowhatter committed:
Bump Zenoh to v1.5.0

Signed-off-by: ChenYing Kuo <evshary@gmail.com>

Conflicts:
	zenoh_cpp_vendor/CMakeLists.txt

yellowhatter committed:
Merge commit 'cdb66eed88a41775e4d6b7a3919805d4963f606b'

Yadunund commented on README.md:
```diff
@@ -226,6 +226,31 @@ Here are two ways to configure on the remote side:
+### Configuration
+
+> [!NOTE]
+> To have Zenoh SHM working, it should be enabled on all Zenoh routers across the message path
```
```suggestion
> To have Zenoh SHM working, it should be enabled on all Zenoh routers across the message path.
```

Yadunund commented on README.md:
````diff
@@ -226,6 +226,31 @@ Here are two ways to configure on the remote side:
+```
+
+The following additional configuration options available as environment variables:
+- `ZENOH_SHM_ALLOC_SIZE`: size (in bytes) of memory to allocate as shared memory arena. Must be a multiply of 4. The default value is 16MB.
+- `ZENOH_SHM_MESSAGE_SIZE_THRESHOLD`: threshold (in bytes) for ROS message wire size to be sent as SHM buffer. Must be a multiply of 4. The default value is 512.
````
```suggestion
- `ZENOH_SHM_ALLOC_SIZE`: size (in bytes) of memory to allocate as shared memory arena. Must be a multiple of 4. The default value is 16MB.
- `ZENOH_SHM_MESSAGE_SIZE_THRESHOLD`: threshold (in bytes) for ROS message wire size to be sent as SHM buffer. Must be a multiple of 4. The default value is 512.
```

Yadunund commented on rmw_zenoh_cpp/src/detail/zenoh_config.hpp:
```diff
@@ -61,6 +61,25 @@ std::optional<zenoh::Config> get_z_config(const ConfigurableEntity & entity);
+/// Get the amount of shared memory to be pre-allocated for Zenoh SHM operation
+/// based on the environment variable ZENOH_SHM_ALLOC_SIZE.
+/// @details The behavior is as follows:
+///   - If not set or <= 0, the default value of 1MB is returned.
```
```suggestion
///   - If not set or <= 0, the default value of 16MB is returned.
```

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -224,19 +225,44 @@ rmw_ret_t PublisherData::publish(
+  if (shm && max_data_length >= shm->msgsize_threshold) {
+    RMW_ZENOH_LOG_DEBUG_NAMED("rmw_zenoh_cpp", "SHM is enabled.");
+
+    auto alloc_result = shm->shm_provider.alloc_gc_defrag(max_data_length);
```
Nit: Replace auto with explicit type here.

Yadunund commented on rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:
```diff
@@ -315,14 +340,52 @@ rmw_ret_t PublisherData::publish_serialized_message(
+
+    auto alloc_result = shm->shm_provider.alloc_gc_defrag(data_length);
+
+    if (std::holds_alternative<zenoh::ZShmMut>(alloc_result)) {
```
`#include <variant>` for `std::holds_alternative`

YuanYuYuan replied:
Resolved in [3f224d7](https://github.com/ros2/rmw_zenoh/pull/363/commits/3f224d749e8c58ca502d2cb60b5044654b13ec70)

Yadunund reviewed (approved):
Thanks for iterating on this PR over months! LGTM from my side but please address comments before running CI and merging.

YuanYuYuan committed:
Apply suggestions from code review

Co-authored-by: yadunund <yadunund@gmail.com>
Signed-off-by: Yuyuan Yuan <az6980522@gmail.com>

YuanYuYuan committed:
chore: address the review

yellowhatter committed:
- fix compilation on windows
- up zenoh-cpp

JEnoch committed:
Merge commit '2477f712aab2ba8204d6d2db3f25c395cf7ef4a1'

Conflicts:
	zenoh_cpp_vendor/CMakeLists.txt

JEnoch commented:
Pulls: ros2/rmw_zenoh#363
Gist: https://gist.githubusercontent.com/JEnoch/eb7066243b40c6606ee411d1ee2fb67f/raw/4cc3ae8525eaadd998d577b813e07a322eee8c19/ros2.repos
BUILD args: "--continue-on-error" --packages-above-and-dependencies zenoh_cpp_vendor zenoh_security_tools rmw_zenoh_cpp
TEST args:  --packages-above zenoh_cpp_vendor zenoh_security_tools rmw_zenoh_cpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16698
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24575)](http://ci.ros2.org/job/ci_linux/24575/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18621)](http://ci.ros2.org/job/ci_linux-aarch64/18621/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4030)](http://ci.ros2.org/job/ci_linux-rhel/4030/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25116)](http://ci.ros2.org/job/ci_windows/25116/)

JEnoch commented:
The failing `ros2doctor.test.test_qos_compatibility` tests on CI are not related to this PR, but to https://github.com/ros2/ros2cli/issues/1088.
